### PR TITLE
Tempo

### PR DESCRIPTION
**Pull request by ireksh**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+clap = { version = "3.1.18", features = ["derive"] }
 anyhow = "1.0"
 rand = "0.8.3"
 serde = { version = "1.0.136", features = ["derive"] }
-serde_json = "1.0.79"
\ No newline at end of file
+serde_json = "1.0.79"
+
+[[example]]
+name = "clap"
```
Comment (damirka) [inline Cargo.toml:16]: Empty line.

File: examples/clap.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+
+use clap::Parser;
```
Comment (damirka) [inline examples/clap.rs:2]: Could skip this empty line here.

File: examples/clap.rs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    count: u8,
+}
+
+fn main() {
+    let args = Args::parse();
+
+    for _ in 0..args.count {
+        println!("Hello {}!", args.name)
+    }
+}
```
Comment (damirka) [inline examples/clap.rs:23]: Do you really need this example? Why is it here?

File: src/main.rs [hunk 4]
```diff
@@ -1,16 +1,104 @@
...
+use anyhow::{Result};
+use clli::{Dictionary, Word};
 use std::{fs::File, io::{prelude::*, BufReader}, path::PathBuf};
 use serde::{Serialize, Deserialize};
 use rand::Rng;
-use structopt::StructOpt;
+use clap::{ Parser, Subcommand};
 
 
 const FILE_PATH: &str = "./examples/storage/turk.json";
```
Comment (damirka) [inline src/main.rs:9]: Path could be specified with a flag. Such as:
```
clli random --path examples/storage/turk.json
```

File: src/main.rs [hunk 5]
```diff
@@ -1,16 +1,104 @@
...
+        },
+
+    /// Get the count of words in the dictionary 
+    Count,
+
+    /// Get the all words from the dictionaryy
+    GetAllWords,
+
+    /// Translate english -> turkish
+    Etranslate{
```
Comment (damirka) [inline src/main.rs:44]: I wouldn't fix on E or T. If we want this solution to be generic, I believe having just one translate would be great. We could imply that most of the time people would translate things from the main language to English. The flow I see here is this:
```
clli translate sozluk
clli translate sozluk --reverse # or perhaps --en meaning that we want to do a translation from the default - en 
```

File: src/main.rs [hunk 6]
```diff
@@ -1,16 +1,104 @@
...
+    Ttranslate{
+        #[clap(required = true)]
+        native: String,
+    },    
+}
 
 fn main() {
-    println!("Hello, sozluk");
+    let args = Cli::parse();
+    let mut sozluk = sozluk_init();
```
Comment (damirka) [inline src/main.rs:58]: Ideally we'd want a dictionary to be set with a custom path:
```
let mut sozluk = sozluk_init(args. ....)
```

File: src/main.rs [hunk 7]
```diff
@@ -1,16 +1,104 @@
...
+            for _ in 0..count {
+                if let Some(word) = sozluk.get_rand_word() {
+                    println!("The random word : {}. translate:{}", &word.native, &word.english );
+                    }
+                }
+            }
+        Commands::Add { native , english } => {
+            println!("Pushing to {} :: {}", native , english);
+            let dummy_cat: Vec<String> = vec![];
+            let dummy_examples: Vec<String> = vec![];
```
Comment (damirka) [inline src/main.rs:71]: Why do we have these guys here? They can be passed into the function directly.

File: src/main.rs [hunk 8]
```diff
@@ -1,16 +1,104 @@
...
+            } 
+        }
+        Commands::Count => {
+            println!("{} records in dictionary", sozluk.count());
+        }
+        Commands::Etranslate {english} => {
+            if let Some(word) = sozluk.get_word(&english) {
+                println!("{} = translate = {} ", word.native, &english);    
+            } else {
+                println!("Translate for {} not found",&english);
```
Comment (damirka) [inline src/main.rs:84]:
```suggestion
                println!("Translation for {} not found",&english);
```

File: src/main.rs [hunk 9]
```diff
@@ -1,16 +1,104 @@
...
+            }
+        }    
+        Commands::Ttranslate {native} => {
+            if let Some(word) = sozluk.get_word_en(&native) {
+                println!("{} = translate = {} ", word.english, &native,);    
+            } else {
+                println!("Translate for {} not found", &native);
```
Comment (damirka) [inline src/main.rs:91]: Same as above

File: src/main.rs [hunk 10]
```diff
@@ -1,16 +1,104 @@
...
+            }
+        }
+        Commands::GetAllWords =>{
+            let mut msg = String::new();
+            for rec in sozluk.get_all().iter(){
+                let _ = &msg.push_str(format!("Turkish: {}, transtlate: {}, examples: {} \n", rec.native, &rec.english, &rec.examples.join("; ")).as_str());
```
Comment (damirka) [inline src/main.rs:97]: Might as well just call println!() multiple times.

**Review by damirka — CHANGES_REQUESTED** [review]
Overall better but few tweaks are needed. We aim to make a generic tool.

File: examples/clap.rs (hunk 3 above)
Comment (ireksh) [reply examples/clap.rs:23]: No need. This code help me to understood  Command enum.

File: src/main.rs (hunk 4 above)
Comment (ireksh) [reply src/main.rs:9]: ok

File: src/main.rs (hunk 7 above)
Comment (ireksh) [reply src/main.rs:71]: Stub. To replace in the future with real vec![String, String, String, String, ].

**Pull request merged**